Expose dist.monitored_barrier() API

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54742 test_c10d: use with_nccl_blocking_wait decorator
* #54741 test_c10d: Run tests with nccl_async_error_handling
* #54740 Provide a decorator to set/unset nccl blocking wait for tests
* #54558 [NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

rohan-varma committed:
Expose dist.monitored_barrier() API

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"


Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

zhaojuanmao reviewed (commented):
overall looks good

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2429,6 +2429,47 @@ def barrier(group=GroupMember.WORLD,
+    Args:
+        group (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+        timeout (float, optional): Timeout in seconds for monitored_barrier.
```
We use `timedelta` for timeout in other APIs like `init_process_group`, should we use `timedelta` here too?

rohan-varma replied:
I see, main reason I did float was for torchscript compatibility and that we did it for RPC framework, but let me check if we can do timedelta and it works with torchscript...

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2429,6 +2429,47 @@ def barrier(group=GroupMember.WORLD,
+        timeout = default_pg_timeout
+
+    group_to_use = _get_default_group() if group is None else group
+    assert isinstance(group_to_use, ProcessGroupGloo)
```
nit: Can remove this since we already check on line 2461?

rohan-varma replied:
Right, get_backend(group) should still do the right check even if group is none.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5120,29 +5120,53 @@ def test_output_unused_in_loss(self):
             # Run monitored barrier
             timeout = 5
-            process_group.monitored_barrier(timeout)
+            dist.monitored_barrier(None, timeout)
```
nit: Use kwarg to avoid `None`: `dist.monitored_barrier(timeout=timeout)`

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5120,29 +5120,53 @@ def test_output_unused_in_loss(self):
+        @require_backends_available({"gloo"})
+        def test_monitored_barrier_gloo_subgroup(self):
+            failed_rank = 1
+            timeout = 5
```
Reduce the timeout to reduce test runtime (1s should be enough).

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"


Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"


Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"


Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"


Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2476,6 +2476,47 @@ def barrier(group=GroupMember.WORLD,
+    will not return from monitored_barrier until all processes have successfully
+    synchronized, if not part of the group, will return immediately.
+
+    This collective always blocks processes until the whole group exits this
```
need to fix the doc here

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5123,52 +5123,67 @@ def test_output_unused_in_loss(self):
+            os.environ["NCCL_BLOCKING_WAIT"] = "1"
+            nccl_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                timeout=timedelta(seconds=10),
```
Can we reduce the timeout here?

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5123,52 +5123,67 @@ def test_output_unused_in_loss(self):
+            if self.rank != 0:
+                with self.assertRaisesRegex(RuntimeError, "Caught collective operation timeout"):
+                    nccl_pg.allreduce(tensors).wait()
                 return
```
Why do we need to do this? Isn't it sufficient to just test that the monitored barrier throws for a subgroup?

rohan-varma replied:
This is weird, I think this commit got messed up due to the hg restack issues internally, and a bunch of the tests went missing. Will send out the correct version

rohan-varma replied:
Yes, I'm not sure why this is rendering in this way but I didn't intend to remove this test. Think something got messed up while rebasing

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5123,52 +5123,67 @@ def test_output_unused_in_loss(self):
-                # out since the barrier is a noop because they are not part of
-                # the subgroup
-                dist.monitored_barrier(subgroup, timeout)
+                err_regex = f"Rank [1-{world_size - 1}]"
```
What exactly is this matching? For world size of 4, this would be "rank [1-3]" but looking at https://github.com/pytorch/pytorch/pull/53773, I don't see any such string in the error we raise? Should the error regex also contain "failed to pass monitoredBarrier"?

rohan-varma replied:
https://github.com/pytorch/pytorch/pull/53773/files#diff-fca17800a316f138000a4cd30902ac0a5fd2b4cd904b562d94e465e5d8255da3R2731 should report "rank x failed to cross barrier" where x is a rank that failed to cross the barrier.

Don't want to put too many constraints on the implementation, so the test just checks that it reported the failure for rank 1, or rank 2, or rank 3 for example, so the regex is rank [1-3]. This is due to iterating the unordered_map, the error can be raised for any rank that did not cross the barrier. Planning to add functionality to report all failed ranks in a follow up change.

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2476,6 +2476,57 @@ def barrier(group=GroupMember.WORLD,
         work.wait()
 
+def monitored_barrier(group=GroupMember.WORLD, timeout=None):
+
```
nit: could explicitly say that this barrier could be inserted before collective calls to check whether some ranks are desynced

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Expose dist.monitored_barrier() API"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* **#53787 Expose dist.monitored_barrier() API**
* #53773 Gloo-only CPU-based monitored barrier

Per title, exposes a python-based monitored barrier API that we can use as part of debugability and may be useful for user applications.

Differential Revision: [D26965127](https://our.internmc.facebook.com/intern/diff/D26965127/)

[ghstack-poisoned]
